### ISSUE TITLE
Fatal Exception: java.lang.IllegalStateException: Reply already submitted

### ISSUE THREAD

**Issue by offline-first** [dooboolab/flutter_inapp_purchase#339]
I got the error from Firebase Crashlytics. Are there any insights into this?  Is this related to #336 @mgonzalezc  ?

```log
Fatal Exception: java.lang.IllegalStateException: Reply already submitted
       at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:35)
       at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:14)
       at com.dooboolab.flutterinapppurchase.MethodResultWrapper$1.run(MethodResultWrapper.java:8)
       at android.os.Handler.handleCallback(Handler.java:873)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:214)
       at android.app.ActivityThread.main(ActivityThread.java:6990)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1445)
```
### Crash on: Android 9, Galaxy Note9 (real device)

- flutter_inapp_purchase: ^5.1.2
- flutter stable channel: 2.5.3
- Android build-tools 31.0.0

**Comment by mgonzalezc** [comment]
Hello! yes, it is

I have ended up using version 5.0.4. The crash isn't there for this version

I think the issue is related to the new handlers in MethodResultWrapper.java, added in #328. From what I could investigate, when getPurchaseHistory is called, the channel receives result.success twice. This is not allowed and that's why "Reply already submitted" fires.

My PR #336 helps a bit because it makes the calls under getPurchaseHistory asynchronous but if you call it several times in your code (for example, with a retry strategy when there is no connection) then the exception is fired as well.

**Comment by offline-first** [comment]
Hello @mgonzalezc 
Thank you for the clarification. I think your PR fix the issue in my case

**Comment by offline-first** [comment]
Ok the problem still exists. I have downgraded to version 5.0.4

**Comment by xalikoutis** [comment]
me too downgraded back to `flutter_inapp_purchase: 5.0.3`

**Comment by 0x-cell** [comment]
I've updated firebase to the latest version, and same error is occurring on  flutter_inapp_purchase: ^5.0.2
```
  firebase_core: ^1.10.6
  firebase_crashlytics: ^2.4.4
  firebase_messaging: ^11.2.4
```

**Comment by jesselpereira** [comment]
Same here.

**Comment by jesselpereira** [comment]
> @jesselpereira I have the same problem :( Did you solve it ??

Hi dude, I just abandoned that package, migrated to **in_app_purchase** and solved my problems.

**Comment by hmarat** [comment]
@jesselpereira okay, thank you for answer :) So in that package it will automatically detect call update purchase callback on app open?))

**Comment by jesselpereira** [comment]
I believe it is worth trying to do this. In my project I have a callback when a purchase is completed, but I don't check it when I open the app. But this is an option that fits my project, nothing prevents you from implementing the package in your project. I believe this package here has been deprecated, or it's not a recognized flaw.

**Comment by giladger** [comment]
Hi,
Any updates on this issue?
It suddenly started happening to me after upgrading to flutter 2.10

**Comment by reju1021** [comment]
> I've updated firebase to the latest version, and same error is occurring on flutter_inapp_purchase: ^5.0.2
> 
> ```
>   firebase_core: ^1.10.6
>   firebase_crashlytics: ^2.4.4
>   firebase_messaging: ^11.2.4
> ```
> 
> It seems that I've found a problem. The `AndroidInappPurchasePlugin.java#onMethodCall` invokes several async operations on the `billingClient.` while persisting the `safeResult` in scope of the entire `AndroidInappPurchasePlugin` instance instead of per `onMethodCall` call. This creates a possibility of having several async tasks call the same instance of `safeResult` instead of the instance of `safeResult` bound to the corresponding `onMethodCall` cycle.
> 
> What i did to fix that was to change the
> 
> ```
> safeResult = new MethodResultWrapper(result, channel);
> ```
> 
> to
> 
> ```
> final MethodResultWrapper safeResult = new MethodResultWrapper(result, channel);
> safeChannel = safeResult;
> ```
> 
> in the `AndroidInappPurchasePlugin.java#onMethodCall`. Note the `safeChannel` - it is used in `private PurchasesUpdatedListener purchasesUpdatedListener = new PurchasesUpdatedListener() {` instead of `safeResult` as it calls the channel anyways.
> 
> Back to drinking. Bye.

Worked like a charm :)

**Comment by giladger** [comment]
@reju1021 @0x-cell can you please share a patch/open a PR?

Thanks!

**Comment by valterh4ck3r** [comment]
Hello Guys.

Use my library with fix

https://github.com/valterh4ck3r/flutter_inapp_purchase

My fix: https://github.com/dooboolab/flutter_inapp_purchase/pull/356

Use this dependency on pubspec.yaml



```
  flutter_inapp_purchase:
    git: https://github.com/valterh4ck3r/flutter_inapp_purchase
```

**Comment by jesselpereira** [comment]
> Hello Guys.
> 
> Use my library with fix
> 
> https://github.com/valterh4ck3r/flutter_inapp_purchase
> 
> My fix: #356
> 
> Use this dependency on pubspec.yaml
> 
> ```
>   flutter_inapp_purchase:
>     git: https://github.com/valterh4ck3r/flutter_inapp_purchase
> ```

Pls pull request your changes. 👏🏻

**Comment by valterh4ck3r** [comment]
Created Pull Request

https://github.com/dooboolab/flutter_inapp_purchase/pull/357

**Comment by Vakil-Parth** [comment]
> > Hello Guys.
> > Use my library with fix
> > https://github.com/valterh4ck3r/flutter_inapp_purchase
> > My fix: #356
> > Use this dependency on pubspec.yaml
> > ```
> >   flutter_inapp_purchase:
> >     git: https://github.com/valterh4ck3r/flutter_inapp_purchase
> > ```
> 
> Pls pull request your changes. 👏🏻

This didn't work for me.

still facing ```java.lang.IllegalStateException: Reply already submitted```

when I call FlutterInappPurchase.instance.getAvailablePurchases() in android

**Comment by offline-first** [comment]
@valterh4ck3r 
Unfortunately the error is still present.
Probably you have to make a purchase first and then call getAvailablePurchases()

**Comment by valterh4ck3r** [comment]
This error happens when flutter calls 2 functions at the same time.

The method call cannot respond at the same time with success or error results, so I've re-implemented the service to not make 2 calls at the same time.

Anyone with the same fix?

**Comment by offline-first** [comment]
@valterh4ck3r 

You have it. Please change the part then it will work
```dart
Future<List<PurchasedItem>?> getPurchaseHistory() async {
    if (_platform.isAndroid) {
      final inappPurchaseHistory = await _channel.invokeMethod(
        'getPurchaseHistoryByType',
        <String, dynamic>{
          'type': describeEnum(_TypeInApp.inapp),
        },
      );

     final subsPurchaseHistory = await _channel.invokeMethod(
        'getPurchaseHistoryByType',
        <String, dynamic>{
          'type': describeEnum(_TypeInApp.subs),
        },
      );

      return extractPurchased(inappPurchaseHistory)! + extractPurchased(subsPurchaseHistory)!;
    } else if (_platform.isIOS) {
      dynamic result = await _channel.invokeMethod('getAvailableItems');

      return extractPurchased(json.encode(result));
    }
    throw PlatformException(
        code: _platform.operatingSystem, message: "platform not supported");
  }
```

**Comment by valterh4ck3r** [comment]
Wait for 1 second, before call method channel again.

```dart
Future<List<PurchasedItem>?> getPurchaseHistory() async {
    if (_platform.isAndroid) {
      final inappPurchaseHistory = await _channel.invokeMethod(
        'getPurchaseHistoryByType',
        <String, dynamic>{
          'type': describeEnum(_TypeInApp.inapp),
        },
      );

   await Future.delayed(Duration(seconds: 1) , (){});

     final subsPurchaseHistory = await _channel.invokeMethod(
        'getPurchaseHistoryByType',
        <String, dynamic>{
          'type': describeEnum(_TypeInApp.subs),
        },
      );

      return extractPurchased(inappPurchaseHistory)! + extractPurchased(subsPurchaseHistory)!;
    } else if (_platform.isIOS) {
      dynamic result = await _channel.invokeMethod('getAvailableItems');

      return extractPurchased(json.encode(result));
    }
    throw PlatformException(
        code: _platform.operatingSystem, message: "platform not supported");
  }
```

**Comment by offline-first** [comment]
The main problem is this line of code
```dart
List<dynamic> results = await Future.wait([getInappPurchaseHistory, getSubsPurchaseHistory]);
```
because the calls are sent in parallel and not in a queue. The timeout of one second is not needed in my test environment. But your fix in Java Source seems to need it too. Can you make a new pull request with my fix which I posted above?

**Comment by valterh4ck3r** [comment]
Yeah, it's perfect.

I will make new pull request, thank you so much @offline-first

**Comment by valterh4ck3r** [comment]
Done.

**Comment by soarescaique** [comment]
@valterh4ck3r @offline-first Any updates on this? I am still facing the same error on version 5.1.2.

**Comment by offline-first** [comment]
@valterh4ck3r please use my fix for the async issue!

**Comment by offline-first** [comment]
@valterh4ck3r I have already pushed my change on your clone
https://github.com/offline-first/flutter_inapp_purchase/commit/228994a0990e8a8409e5a10794d51c49595eb34e#diff-09ef645a93240c5c6f0a3209c0af6a335d2cdb16bf552e1e5533677bee2e5202

**Comment by offline-first** [comment]
@valterh4ck3r done

**Comment by valterh4ck3r** [comment]
Thank you so much

**Comment by valterh4ck3r** [comment]
@offline-first merged.

**Comment by soarescaique** [comment]
@valterh4ck3r @offline-first Sorry if I am getting it wrong. But why was the PR #359 closed since it has already been approved to be merged by @hyochan. Thank you guys for the collaboration on the package.

**Comment by hyochan** [comment]
I think @valterh4ck3r has trouble managing the branch. I'll open up the `PR` again since it looks like the [latest commit](https://github.com/dooboolab/flutter_inapp_purchase/pull/359/commits/eb3dc9534e792382615899e6cb05bcd72dbd5c60) was intended.

**Comment by hyochan** [comment]
Closes due to #360 

Released in `5.2.0`

**Comment by offline-first** [comment]
Thanks @hyochan, purchase history is now works correctly again on android
